server/lib/model/action.dart

```diff
@@ -0,0 +1,73 @@
+import 'dart:async';
+
+import 'package:fixnum/fixnum.dart';
+
+import '../proto/rule.pb.dart' as proto;
+import '../service/dv360.dart';
+
+/// An interface that represents an action to manipulate DV360 entities.
+abstract class Action {
+  /// Creates an instance of a class that implements [Action], from a proto.
+  ///
+  /// Throws an [UnsupportedError] if the proto's action type is unspecified.
+  factory Action.fromProto(proto.Action action) {
```

Comment:
What is the advantage of using factory here? Would creating a concrete method will serve the purpose? We can also use general constructor in other files `ChangeLineItemStatusAction(proto.Action action)` so then our code does not look as complicated.

```
getActionFromProto(proto.Action action) {
   switch (action.type) {
      case proto.Action_Type.CHANGE_LINE_ITEM_STATUS:
        return ChangeLineItemStatusAction(action);
      default:
        throw UnsupportedError(
            '${action.type} is not a supported action type.');
}
```

Reply:
That makes sense. There's no inherent advantage to using `factory`, I guess I was just trying to avoid the need to call a constructor then a method. But you're right, that makes the code less complicated.

Reply:
In dart, you can implement a method in abstract class and the subclass inherit the implementation so then you don't need to reimplement in each subclass. With the `Action` in `Action.getActionFromProto(...)`, it is hard to track where method, "getActionFromProto()", comes from in the caller class. 

Search 'concrete methods' in
https://www.learndartprogramming.com/fundamentals/abstract-classes-in-dart/

Reply:
I get `error: Missing concrete implementation of 'Action.getActionFromProto'` when I try to create a concrete method.

I think for subclasses that `extends` a class, you can inherit the concrete method, but with subtypes that `implements` an interface, you have to provide an implementation for every method in the interface.

Should I provide a `'noSuchMethod' implementation` for subtypes, or maybe use inheritance?

Reply:
Actually, I guess `factory` does have a use here. Because, if we change this method to a factory, it becomes a constructor that can return subtypes, and since the constructor isn't a part of the interface, it doesn't have to implemented in subtypes.

Reply:
I thought you created an abstract class so then the subclasses and `extends` from it. But seems like you use the abstract class as an interface. Yes, if you use it as an interface, you will need to implement all the methods. Have you considered `extends` Action class instead of `implements` it so then you don't need to reimplement the `getActionFromProto()` in every subclass? Or you expect the subclasses will inherit more interfaces? A class can inherit one abstract class and implement multiple interfaces.

Reply:
Hm, I wasn't sure if it made sense to allow `changeLineItemStatusAction.getActionFromProto(protoAction)` if I use `extends`.

How about making `getActionFromProto()` a static method?

Reply:
Sounds good. As we discuss offline, @thu5 will implement a static method in `action.dar`t so the caller class call `Action.getActionFromProto()`